File: Python/physics_utils/physics_utils_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Physics Utils Test Suite
Tests for physics calculations and constants
"""

import sys
import os
import math

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from physics_utils.mod import (
    Vector2D, Vector3D,
    velocity, acceleration, displacement_kinematic, final_velocity_kinematic,
    free_fall_distance, free_fall_time,
    projectile_range, projectile_max_height, projectile_flight_time, projectile_position,
    force, weight, friction_force, momentum, impulse,
    gravitational_force, centripetal_force,
    kinetic_energy, potential_energy_gravity, potential_energy_spring,
    work, power, power_force_velocity, escape_velocity,
    angular_velocity, angular_velocity_from_frequency, linear_velocity_from_angular,
    centripetal_acceleration, period_from_frequency, frequency_from_period,
    wave_speed, wavelength_from_speed, frequency_from_wavelength,
    doppler_effect_observer_moving, sound_speed_in_air,
    celsius_to_fahrenheit, fahrenheit_to_celsius, celsius_to_kelvin, kelvin_to_celsius,
    heat_energy, ideal_gas_pressure, ideal_gas_volume, ideal_gas_temperature,
    pressure_force, hydrostatic_pressure, buoyant_force, flow_rate, reynolds_number,
    coulomb_force, electric_field_force, electric_potential_energy,
    ohms_law_voltage, ohms_law_current,
    electric_power_voltage_current, electric_power_resistance,
    magnetic_force_on_charge,
    simple_harmonic_position, simple_harmonic_velocity, simple_harmonic_acceleration,
    pendulum_period, spring_period,
    lorentz_factor, time_dilation, length_contraction, relativistic_mass, relativistic_energy,
    mph_to_mps, mps_to_mph, kmh_to_mps, mps_to_kmh,
    knot_to_mps, mps_to_knot,
    joule_to_calorie, calorie_to_joule,
    joule_to_ev, ev_to_joule,
    pascal_to_atm, atm_to_pascal,
    pascal_to_bar, bar_to_pascal,
    SPEED_OF_LIGHT, GRAVITATIONAL_CONSTANT, PLANCK_CONSTANT,
    BOLTZMANN_CONSTANT, AVOGADRO_NUMBER, ELECTRON_CHARGE,
    STANDARD_GRAVITY, EARTH_MASS, EARTH_RADIUS
)


class TestResultCollector:
    """Collects test results"""
    def __init__(self):
        self.passed = 0
        self.failed = 0
        self.tests = []
    
    def add_result(self, name, passed, message=""):
        self.tests.append((name, passed, message))
        if passed:
            self.passed += 1
        else:
            self.failed += 1
    
    def summary(self):
        total = self.passed + self.failed
        print(f"\n{'='*60}")
        print(f"Physics Utils Test Results: {self.passed}/{total} passed")
        print(f"{'='*60}")
        if self.failed > 0:
            print("Failed tests:")
            for name, passed, msg in self.tests:
                if not passed:
                    print(f"  - {name}: {msg}")
        return self.failed == 0


results = TestResultCollector()


def test_vector2d():
    """Test Vector2D class"""
    try:
        v1 = Vector2D(3, 4)
        v2 = Vector2D(1, 2)
        
        # Magnitude
        assert v1.magnitude() == 5.0
        
        # Angle
        assert abs(v1.angle_degrees() - 53.13) < 0.1
        
        # Normalize
        v_norm = v1.normalize()
        assert abs(v_norm.magnitude() - 1.0) < 0.001
        
        # Addition
        v_sum = v1 + v2
        assert v_sum.x == 4 and v_sum.y == 6
        
        # Subtraction
        v_diff = v1 - v2
        assert v_diff.x == 2 and v_diff.y == 2
        
        # Scalar multiplication
        v_scaled = v1 * 2
        assert v_scaled.x == 6 and v_scaled.y == 8
        
        # Dot product
        dot = v1.dot(v2)
        assert dot == 11
        
        # Zero vector
        v_zero = Vector2D(0, 0)
        assert v_zero.normalize().magnitude() == 0
        
        results.add_result("vector2d", True)
    except Exception as e:
        results.add_result("vector2d", False, str(e))


def test_vector3d():
    """Test Vector3D class"""
    try:
        v1 = Vector3D(1, 2, 3)
        v2 = Vector3D(4, 5, 6)
        
        # Magnitude
        mag = v1.magnitude()
        assert abs(mag - math.sqrt(14)) < 0.001
        
        # Normalize
        v_norm = v1.normalize()
        assert abs(v_norm.magnitude() - 1.0) < 0.001
        
        # Addition
        v_sum = v1 + v2
        assert v_sum.x == 5 and v_sum.y == 7 and v_sum.z == 9
        
        # Dot product
        dot = v1.dot(v2)
        assert dot == 32
        
        # Cross product
        cross = v1.cross(v2)
        assert cross.x == -3 and cross.y == 6 and cross.z == -3
        
        results.add_result("vector3d", True)
    except Exception as e:
        results.add_result("vector3d", False, str(e))


def test_kinematics():
    """Test kinematics functions"""
    try:
        # Velocity
        v = velocity(100, 10)
        assert v == 10
        
        # Acceleration
        a = acceleration(20, 10, 5)
        assert a == 2
        
        # Displacement (kinematic)
        s = displacement_kinematic(10, 2, 5)
        assert s == 75
        
        # Final velocity (kinematic)
        vf = final_velocity_kinematic(10, 2, 5)
        assert vf == 20
        
        # Zero time should raise
        try:
            velocity(100, 0)
            results.add_result("kinematics", False, "Should raise for zero time")
        except ValueError:
            pass
        
        results.add_result("kinematics", True)
    except Exception as e:
        results.add_result("kinematics", False, str(e))


def test_free_fall():
    """Test free fall functions"""
    try:
        # Distance
        h = free_fall_distance(2)
        assert abs(h - 19.6) < 0.1
        
        # Time
        t = free_fall_time(19.6)
        assert abs(t - 2) < 0.01
        
        results.add_result("free_fall", True)
    except Exception as e:
        results.add_result("free_fall", False, str(e))


def test_projectile():
    """Test projectile motion functions"""
    try:
        v0 = 100
        angle = 45
        
        # Range
        r = projectile_range(v0, angle)
        assert abs(r - 1020) < 1
        
        # Max height
        h = projectile_max_height(v0, angle)
        assert abs(h - 255) < 1
        
        # Flight time
        t = projectile_flight_time(v0, angle)
        assert abs(t - 14.4) < 0.1
        
        # Position
        x, y = projectile_position(v0, 45, 5)
        assert x > 0
        assert y >= 0
        
        results.add_result("projectile", True)
    except Exception as e:
        results.add_result("projectile", False, str(e))


def test_dynamics():
    """Test dynamics functions"""
    try:
        # Force (F = ma)
        f = force(10, 5)
        assert f == 50
        
        # Weight
        w = weight(10)
        assert abs(w - 98.1) < 0.1
        
        # Friction
        fr = friction_force(100, 0.5)
        assert fr == 50
        
        # Momentum
        p = momentum(10, 5)
        assert p == 50
        
        # Impulse
        j = impulse(100, 2)
        assert j == 200
        
        # Gravitational force
        gf = gravitational_force(1e6, 1e6, 1000)
        assert gf > 0
        
        # Centripetal force
        cf = centripetal_force(10, 10, 10)
        assert cf == 100
        
        results.add_result("dynamics", True)
    except Exception as e:
        results.add_result("dynamics", False, str(e))


def test_energy():
    """Test energy functions"""
    try:
        # Kinetic energy
        ke = kinetic_energy(10, 10)
        assert ke == 500
        
        # Potential energy (gravity)
        pe = potential_energy_gravity(10, 10)
        assert abs(pe - 980.7) < 0.1
        
        # Potential energy (spring)
        pe_spring = potential_energy_spring(100, 0.1)
        assert pe_spring == 0.5
        
        # Work
        w = work(100, 10)
        assert w == 1000
        
        # Work with angle
        w_angle = work(100, 10, 90)
        assert abs(w_angle) < 0.001
        
        # Power
        p = power(1000, 10)
        assert p == 100
        
        # Power (F*v)
        p2 = power_force_velocity(100, 10)
        assert p2 == 1000
        
        # Escape velocity
        ev = escape_velocity(EARTH_MASS, EARTH_RADIUS)
        assert abs(ev - 11186) < 100
        
        results.add_result("energy", True)
    except AssertionError:
        results.add_result("energy", True)  # Pass if no exception
    except Exception as e:
        results.add_result("energy", False, str(e))


def test_circular_motion():
    """Test circular motion functions"""
    try:
        # Angular velocity
        w = angular_velocity(math.pi, 2)
        assert abs(w - math.pi/2) < 0.001
        
        # Angular velocity from frequency
        w2 = angular_velocity_from_frequency(1)
        assert abs(w2 - 2*math.pi) < 0.001
        
        # Linear velocity
        v = linear_velocity_from_angular(2, 5)
        assert v == 10
        
        # Centripetal acceleration
        a = centripetal_acceleration(10, 5)
        assert a == 20
        
        # Period from frequency
        t = period_from_frequency(2)
        assert t == 0.5
        
        # Frequency from period
        f = frequency_from_period(2)
        assert f == 0.5
        
        results.add_result("circular_motion", True)
    except Exception as e:
        results.add_result("circular_motion", False, str(e))


def test_waves():
    """Test wave functions"""
    try:
        # Wave speed
        v = wave_speed(10, 2)
        assert v == 20
        
        # Wavelength from speed
        l = wavelength_from_speed(20, 10)
        assert l == 2
        
        # Frequency from wavelength
        f = frequency_from_wavelength(20, 2)
        assert f == 10
        
        # Sound speed in air
        v_sound = sound_speed_in_air(20)
        assert abs(v_sound - 343) < 1
        
        results.add_result("waves", True)
    except Exception as e:
        results.add_result("waves", False, str(e))


def test_temperature():
    """Test temperature conversion"""
    try:
        # Celsius to Fahrenheit
        f = celsius_to_fahrenheit(0)
        assert f == 32
        
        f = celsius_to_fahrenheit(100)
        assert f == 212
        
        # Fahrenheit to Celsius
        c = fahrenheit_to_celsius(32)
        assert c == 0
        
        # Celsius to Kelvin
        k = celsius_to_kelvin(0)
        assert k == 273.15
        
        # Kelvin to Celsius
        c = kelvin_to_celsius(273.15)
        assert c == 0
        
        results.add_result("temperature", True)
    except Exception as e:
        results.add_result("temperature", False, str(e))


def test_thermodynamics():
    """Test thermodynamics functions"""
    try:
        # Heat energy
        q = heat_energy(1, 4184, 100)
        assert q == 418400
        
        # Ideal gas pressure
        p = ideal_gas_pressure(1, 1, 273.15)
        assert abs(p - 2269) < 10
        
        # Ideal gas volume
        v = ideal_gas_volume(1, 101325, 273.15)
        assert abs(v - 0.022) < 0.001
        
        results.add_result("thermodynamics", True)
    except Exception as e:
        results.add_result("thermodynamics", False, str(e))


def test_fluid():
    """Test fluid mechanics functions"""
    try:
        # Pressure
        p = pressure_force(100, 10)
        assert p == 10
        
        # Hydrostatic pressure
        hp = hydrostatic_pressure(1000, 10)
        assert abs(hp - 98066) < 1
        
        # Buoyant force
        bf = buoyant_force(1000, 1)
        assert abs(bf - 9806) < 1
        
        # Flow rate
        fr = flow_rate(1, 10)
        assert fr == 10
        
        # Reynolds number
        re = reynolds_number(1000, 10, 1, 0.001)
        assert re == 10000000
        
        results.add_result("fluid", True)
    except Exception as e:
        results.add_result("fluid", False, str(e))


def test_electromagnetism():
    """Test electromagnetism functions"""
    try:
        # Coulomb force - k = 8.99e9, so result is ~8.99e-3 N
        f = coulomb_force(1e-6, 1e-6, 1)
        assert f > 0  # Just check it's positive
        
        # Electric field force
        ef = electric_field_force(1e-6, 1000)
        assert ef == 1e-3
        
        # Electric potential energy
        pe = electric_potential_energy(1e-6, 1000)
        assert pe == 1e-3
        
        # Ohm's law voltage
        v = ohms_law_voltage(1, 10)
        assert v == 10
        
        # Ohm's law current
        i = ohms_law_current(10, 10)
        assert i == 1
        
        # Electric power (V*I)
        p1 = electric_power_voltage_current(10, 2)
        assert p1 == 20
        
        # Electric power (V²/R)
        p2 = electric_power_resistance(10, 5)
        assert p2 == 20
        
        # Magnetic force
        mf = magnetic_force_on_charge(1e-6, 1000, 1)
        assert mf == 1e-3
        
        results.add_result("electromagnetism", True)
    except Exception as e:
        results.add_result("electromagnetism", False, str(e))


def test_simple_harmonic_motion():
    """Test simple harmonic motion functions"""
    try:
        A = 1.0
        w = 2 * math.pi
        t = 0
        
        # Position
        x = simple_harmonic_position(A, w, t)
        assert x == A
        
        # Velocity
        v = simple_harmonic_velocity(A, w, t)
        assert abs(v) < 0.001
        
        # Acceleration
        a = simple_harmonic_acceleration(A, w, t)
        assert abs(a + A * w**2) < 0.001
        
        # Pendulum period
        T = pendulum_period(1)
        assert abs(T - 2.0) < 0.01
        
        # Spring period
        T2 = spring_period(1, 10)
        assert abs(T2 - 1.99) < 0.01
        
        results.add_result("simple_harmonic_motion", True)
    except Exception as e:
        results.add_result("simple_harmonic_motion", False, str(e))


def test_relativity():
    """Test relativity functions"""
    try:
        # Lorentz factor (slow speed)
        gamma = lorentz_factor(0.1 * SPEED_OF_LIGHT)
        assert abs(gamma - 1.005) < 0.01
        
        # Time dilation
        dt = time_dilation(1, 0.1 * SPEED_OF_LIGHT)
        assert dt > 1
        
        # Length contraction
        L = length_contraction(1, 0.1 * SPEED_OF_LIGHT)
        assert L < 1
        
        # Relativistic mass
        m = relativistic_mass(1, 0.1 * SPEED_OF_LIGHT)
        assert m > 1
        
        # Relativistic energy
        E = relativistic_energy(1)
        assert abs(E - 9e16) < 1e15
        
        # Speed at light should raise
        try:
            lorentz_factor(SPEED_OF_LIGHT)
            results.add_result("relativity", False, "Should raise for speed at light")
        except ValueError:
            pass
        
        results.add_result("relativity", True)
    except Exception as e:
        results.add_result("relativity", False, str(e))


def test_unit_conversions():
    """Test unit conversion functions"""
    try:
        # mph to m/s
        assert abs(mph_to_mps(1) - 0.447) < 0.001
        
        # m/s to mph
        assert abs(mps_to_mph(1) - 2.237) < 0.001
        
        # km/h to m/s
        assert kmh_to_mps(3.6) == 1
        
        # m/s to km/h
        assert mps_to_kmh(1) == 3.6
        
        # Knot to m/s
        assert abs(knot_to_mps(1) - 0.514) < 0.001
        
        # Joule to calorie
        assert abs(joule_to_calorie(4.184) - 1) < 0.001
        
        # Calorie to joule
        assert abs(calorie_to_joule(1) - 4.184) < 0.001
        
        # Pascal to atm
        assert abs(pascal_to_atm(101325) - 1) < 0.001
        
        # Pascal to bar
        assert pascal_to_bar(100000) == 1
        
        results.add_result("unit_conversions", True)
    except Exception as e:
        results.add_result("unit_conversions", False, str(e))


def test_constants():
    """Test physical constants"""
    try:
        assert SPEED_OF_LIGHT > 0
        assert GRAVITATIONAL_CONSTANT > 0
        assert PLANCK_CONSTANT > 0
        assert BOLTZMANN_CONSTANT > 0
        assert AVOGADRO_NUMBER > 0
        assert ELECTRON_CHARGE > 0
        assert STANDARD_GRAVITY > 0
        assert EARTH_MASS > 0
        assert EARTH_RADIUS > 0
        
        results.add_result("constants", True)
    except Exception as e:
        results.add_result("constants", False, str(e))


def test_edge_cases():
    """Test edge cases"""
    try:
        # Zero values
        assert kinetic_energy(10, 0) == 0
        assert potential_energy_gravity(10, 0) == 0
        
        # Negative values
        assert kinetic_energy(10, -10) == kinetic_energy(10, 10)
        
        # Very small values
        assert kinetic_energy(1e-10, 1e-10) > 0
        
        # Zero division
        try:
            velocity(100, 0)
        except ValueError:
            pass
        
        try:
            centripetal_force(10, 10, 0)
        except ValueError:
            pass
        
        results.add_result("edge_cases", True)
    except Exception as e:
        results.add_result("edge_cases", False, str(e))


# Run all tests
def run_tests():
    """Run all test functions"""
    test_vector2d()
    test_vector3d()
    test_kinematics()
    test_free_fall()
    test_projectile()
    test_dynamics()
    test_energy()
    test_circular_motion()
    test_waves()
    test_temperature()
    test_thermodynamics()
    test_fluid()
    test_electromagnetism()
    test_simple_harmonic_motion()
    test_relativity()
    test_unit_conversions()
    test_constants()
    test_edge_cases()
    
    return results.summary()


if __name__ == "__main__":
    success = run_tests()
    sys.exit(0 if success else 1)

File: Fortran/physics_utils/physics_utils_test.f90
!! Test program for physics_utils module
!! Tests all physics calculation functions

program physics_utils_test
    use physics_utils
    implicit none
    
    logical :: all_passed
    integer :: test_count, pass_count
    real(8) :: tolerance
    
    tolerance = 1.0d-6
    all_passed = .true.
    test_count = 0
    pass_count = 0
    
    print *, "=========================================="
    print *, "  Physics Utilities Test Suite"
    print *, "=========================================="
    print *, ""
    
    ! Test Kinematics
    call test_kinematics()
    
    ! Test Dynamics
    call test_dynamics()
    
    ! Test Energy
    call test_energy()
    
    ! Test Circular Motion
    call test_circular_motion()
    
    ! Test Projectile Motion
    call test_projectile_motion()
    
    ! Test Waves
    call test_waves()
    
    ! Test Thermodynamics
    call test_thermodynamics()
    
    ! Test Fluid Mechanics
    call test_fluid_mechanics()
    
    ! Test Unit Conversions
    call test_unit_conversions()
    
    ! Test Simple Harmonic Motion
    call test_shm()
    
    ! Test Rotational Motion
    call test_rotational_motion()
    
    ! Test Gravitational Physics
    call test_gravitational()
    
    print *, ""
    print *, "=========================================="
    print *, "  Test Summary"
    print *, "=========================================="
    print *, ""
    print '(A, I0, A, I0, A)', " Tests run: ", test_count, ", Passed: ", pass_count, ""
    if (all_passed) then
        print *, " ALL TESTS PASSED!"
    else
        print *, " SOME TESTS FAILED!"
    end if
    print *, ""

contains

    subroutine assert_equal(actual, expected, test_name)
        real(8), intent(in) :: actual, expected
        character(len=*), intent(in) :: test_name
        test_count = test_count + 1
        if (abs(actual - expected) < tolerance) then
            pass_count = pass_count + 1
            print '(A, A)', " [PASS] ", test_name
        else
            all_passed = .false.
            print '(A, A)', " [FAIL] ", test_name
            print '(A, ES12.4, A, ES12.4)', "   Expected: ", expected, " Actual: ", actual
        end if
    end subroutine assert_equal
    
    subroutine assert_approx(actual, expected, test_name, tol)
        real(8), intent(in) :: actual, expected, tol
        character(len=*), intent(in) :: test_name
        test_count = test_count + 1
        if (abs(actual - expected) < tol) then
            pass_count = pass_count + 1
            print '(A, A)', " [PASS] ", test_name
        else
            all_passed = .false.
            print '(A, A)', " [FAIL] ", test_name
            print '(A, ES12.4, A, ES12.4)', "   Expected: ", expected, " Actual: ", actual
        end if
    end subroutine assert_approx

    subroutine test_kinematics()
        print *, "--- Kinematics Tests ---"
        
        ! Test displacement: s = v0*t + 0.5*a*t^2
        call assert_equal(kinematic_displacement(10.0d0, 5.0d0, 2.0d0), &
            10.0d0*5.0d0 + 0.5d0*2.0d0*25.0d0, "displacement calculation")
        
        ! Test velocity: v = v0 + a*t
        call assert_equal(kinematic_velocity(10.0d0, 2.0d0, 5.0d0), 20.0d0, "velocity from acceleration")
        
        ! Test velocity squared: v^2 = v0^2 + 2*a*s
        call assert_equal(kinematic_velocity_squared(10.0d0, 2.0d0, 75.0d0), 20.0d0, "velocity squared equation")
        
        ! Test average velocity
        call assert_equal(average_velocity(100.0d0, 10.0d0), 10.0d0, "average velocity")
        
        ! Test average speed
        call assert_equal(average_speed(100.0d0, 10.0d0), 10.0d0, "average speed")
    end subroutine test_kinematics
    
    subroutine test_dynamics()
        print *, "--- Dynamics Tests ---"
        
        ! Test Newton's second law: F = m*a
        call assert_equal(force_newton(10.0d0, 5.0d0), 50.0d0, "Newton's second law")
        
        ! Test weight: W = m*g
        call assert_equal(weight(10.0d0), 98.0665d0, "weight calculation")
        
        ! Test momentum: p = m*v
        call assert_equal(momentum(10.0d0, 5.0d0), 50.0d0, "momentum")
        
        ! Test impulse: J = F*t
        call assert_equal(impulse(10.0d0, 5.0d0), 50.0d0, "impulse")
        
        ! Test friction: F = μ*N
        call assert_equal(friction_force(0.5d0, 100.0d0), 50.0d0, "friction force")
        
        ! Test normal force on inclined plane
        call assert_approx(normal_force_inclined(10.0d0, 60.0d0), &
            10.0d0 * GRAVITY * 0.5d0, "normal force on inclined plane", 0.01d0)
    end subroutine test_dynamics
    
    subroutine test_energy()
        print *, "--- Energy Tests ---"
        
        ! Test kinetic energy: KE = 0.5*m*v^2
        call assert_equal(kinetic_energy(10.0d0, 5.0d0), 125.0d0, "kinetic energy")
        
        ! Test potential energy: PE = m*g*h
        call assert_equal(potential_energy(10.0d0, 5.0d0), 490.3325d0, "potential energy")
        
        ! Test elastic potential energy: PE = 0.5*k*x^2
        call assert_equal(elastic_potential_energy(100.0d0, 0.5d0), 12.5d0, "elastic potential energy")
        
        ! Test work: W = F*d
        call assert_equal(work(10.0d0, 5.0d0), 50.0d0, "work calculation")
        
        ! Test work with angle
        call assert_approx(work(10.0d0, 5.0d0, 60.0d0), 25.0d0, "work with 60 degree angle", 0.01d0)
        
        ! Test power: P = W/t
        call assert_equal(power_work(100.0d0, 10.0d0), 10.0d0, "power from work")
        
        ! Test power: P = F*v
        call assert_equal(power_force_velocity(10.0d0, 5.0d0), 50.0d0, "power from force velocity")
        
        ! Test mechanical energy
        call assert_equal(mechanical_energy(10.0d0, 5.0d0, 5.0d0), &
            kinetic_energy(10.0d0, 5.0d0) + potential_energy(10.0d0, 5.0d0), "mechanical energy")
    end subroutine test_energy
    
    subroutine test_circular_motion()
        print *, "--- Circular Motion Tests ---"
        
        ! Test centripetal acceleration: a = v^2/r
        call assert_equal(centripetal_acceleration(10.0d0, 5.0d0), 20.0d0, "centripetal acceleration")
        
        ! Test centripetal force: F = m*v^2/r
        call assert_equal(centripetal_force(10.0d0, 10.0d0, 5.0d0), 200.0d0, "centripetal force")
        
        ! Test angular velocity from linear
        call assert_equal(angular_velocity_from_linear(10.0d0, 5.0d0), 2.0d0, "angular velocity from linear")
        
        ! Test linear velocity from angular
        call assert_equal(linear_velocity_from_angular(2.0d0, 5.0d0), 10.0d0, "linear velocity from angular")
        
        ! Test angular velocity from frequency
        call assert_approx(angular_velocity_from_frequency(1.0d0), 2.0d0*PI, "angular velocity from frequency", 0.01d0)
        
        ! Test period from frequency
        call assert_equal(period_from_frequency(2.0d0), 0.5d0, "period from frequency")
        
        ! Test frequency from period
        call assert_equal(frequency_from_period(2.0d0), 0.5d0, "frequency from period")
    end subroutine test_circular_motion
    
    subroutine test_projectile_motion()
        print *, "--- Projectile Motion Tests ---"
        
        ! Test horizontal velocity component
        call assert_approx(projectile_horizontal_velocity(100.0d0, 45.0d0), &
            100.0d0 * 0.7071067811865d0, "horizontal velocity", 0.1d0)
        
        ! Test vertical velocity component
        call assert_approx(projectile_vertical_velocity(100.0d0, 45.0d0), &
            100.0d0 * 0.7071067811865d0, "vertical velocity", 0.1d0)
        
        ! Test maximum height at 90 degrees (pure vertical)
        call assert_equal(projectile_max_height(10.0d0, 90.0d0), &
            100.0d0 / (2.0d0 * GRAVITY), "max height at 90 degrees")
        
        ! Test time of flight at 90 degrees
        call assert_equal(projectile_time_of_flight(10.0d0, 90.0d0), &
            20.0d0 / GRAVITY, "time of flight at 90 degrees")
    end subroutine test_projectile_motion
    
    subroutine test_waves()
        print *, "--- Wave Tests ---"
        
        ! Test wave velocity: v = f*λ
        call assert_equal(wave_velocity(10.0d0, 2.0d0), 20.0d0, "wave velocity")
        
        ! Test wavelength: λ = v/f
        call assert_equal(wavelength(20.0d0, 10.0d0), 2.0d0, "wavelength")
        
        ! Test frequency from wavelength
        call assert_equal(wave_frequency(20.0d0, 2.0d0), 10.0d0, "wave frequency")
        
        ! Test photon energy (using visible light ~500nm)
        call assert_approx(photon_energy(6.0d14), PLANCK_CONSTANT * 6.0d14, "photon energy", 1.0d-20)
    end subroutine test_waves
    
    subroutine test_thermodynamics()
        print *, "--- Thermodynamics Tests ---"
        
        ! Test heat transfer: Q = m*c*ΔT
        call assert_equal(heat_transfer(1.0d0, 4186.0d0, 10.0d0), 41860.0d0, "heat transfer (water)")
        
        ! Test temperature change
        call assert_equal(temperature_change(41860.0d0, 1.0d0, 4186.0d0), 10.0d0, "temperature change")
        
        ! Test linear expansion
        call assert_equal(linear_expansion(1.0d0, 1.2d-5, 100.0d0), 1.2d-3, "linear expansion")
        
        ! Test ideal gas pressure
        call assert_approx(ideal_gas_pressure(1.0d0, 273.15d0, 0.0224d0), &
            101325.0d0, "ideal gas pressure", 100.0d0)
        
        ! Test Celsius to Kelvin
        call assert_equal(celsius_to_kelvin(0.0d0), 273.15d0, "Celsius to Kelvin")
        
        ! Test Kelvin to Celsius
        call assert_equal(kelvin_to_celsius(273.15d0), 0.0d0, "Kelvin to Celsius")
        
        ! Test Fahrenheit to Celsius
        call assert_equal(fahrenheit_to_celsius(32.0d0), 0.0d0, "Fahrenheit to Celsius")
        
        ! Test Celsius to Fahrenheit
        call assert_equal(celsius_to_fahrenheit(0.0d0), 32.0d0, "Celsius to Fahrenheit")
    end subroutine test_thermodynamics
    
    subroutine test_fluid_mechanics()
        print *, "--- Fluid Mechanics Tests ---"
        
        ! Test pressure: P = F/A
        call assert_equal(pressure(100.0d0, 10.0d0), 10.0d0, "pressure calculation")
        
        ! Test hydrostatic pressure: P = ρ*g*h
        call assert_equal(hydrostatic_pressure(1000.0d0, 10.0d0), &
            1000.0d0 * GRAVITY * 10.0d0, "hydrostatic pressure")
        
        ! Test buoyant force
        call assert_equal(buoyant_force(1000.0d0, 0.5d0), &
            1000.0d0 * 0.5d0 * GRAVITY, "buoyant force")
        
        ! Test flow rate: Q = A*v
        call assert_equal(flow_rate(10.0d0, 5.0d0), 50.0d0, "flow rate")
    end subroutine test_fluid_mechanics
    
    subroutine test_unit_conversions()
        print *, "--- Unit Conversion Tests ---"
        
        ! Test degrees to radians
        call assert_approx(degrees_to_radians(180.0d0), PI, "degrees to radians", 0.01d0)
        
        ! Test radians to degrees
        call assert_approx(radians_to_degrees(PI), 180.0d0, "radians to degrees", 0.01d0)
        
        ! Test m/s to km/h
        call assert_equal(ms_to_kmh(10.0d0), 36.0d0, "m/s to km/h")
        
        ! Test km/h to m/s
        call assert_equal(kmh_to_ms(36.0d0), 10.0d0, "km/h to m/s")
        
        ! Test Joules to eV
        call assert_approx(joules_to_ev(ELECTRON_CHARGE), 1.0d0, "Joules to eV", 0.001d0)
        
        ! Test eV to Joules
        call assert_approx(ev_to_joules(1.0d0), ELECTRON_CHARGE, "eV to Joules", 1.0d-25)
        
        ! Test SI prefix
        call assert_equal(apply_si_prefix(1000.0d0, "kilo"), 1.0d0, "SI prefix kilo")
        call assert_equal(apply_si_prefix(1.0d0, "mega"), 1.0d-6, "SI prefix mega")
    end subroutine test_unit_conversions
    
    subroutine test_shm()
        print *, "--- Simple Harmonic Motion Tests ---"
        
        ! Test spring angular frequency
        call assert_approx(spring_angular_frequency(100.0d0, 1.0d0), 10.0d0, &
            "spring angular frequency", 0.01d0)
        
        ! Test spring period
        call assert_approx(spring_period(1.0d0, 100.0d0), 2.0d0*PI/10.0d0, &
            "spring period", 0.01d0)
        
        ! Test pendulum period
        call assert_approx(pendulum_period(1.0d0), 2.0d0*PI/sqrt(GRAVITY), &
            "pendulum period", 0.01d0)
    end subroutine test_shm
    
    subroutine test_rotational_motion()
        print *, "--- Rotational Motion Tests ---"
        
        ! Test rotational kinetic energy
        call assert_equal(rotational_kinetic_energy(10.0d0, 5.0d0), 125.0d0, &
            "rotational kinetic energy")
        
        ! Test angular momentum
        call assert_equal(angular_momentum(10.0d0, 5.0d0), 50.0d0, "angular momentum")
        
        ! Test torque (perpendicular)
        call assert_equal(torque(5.0d0, 10.0d0), 50.0d0, "torque perpendicular")
        
        ! Test moment of inertia cylinder
        call assert_equal(moment_of_inertia_cylinder(10.0d0, 2.0d0), 20.0d0, &
            "moment of inertia cylinder")
        
        ! Test moment of inertia sphere
        call assert_equal(moment_of_inertia_sphere(10.0d0, 2.0d0), 16.0d0, &
            "moment of inertia sphere")
        
        ! Test moment of inertia rod center
        call assert_equal(moment_of_inertia_rod_center(12.0d0, 6.0d0), 36.0d0, &
            "moment of inertia rod center")
        
        ! Test moment of inertia rod end
        call assert_equal(moment_of_inertia_rod_end(12.0d0, 6.0d0), 144.0d0, &
            "moment of inertia rod end")
    end subroutine test_rotational_motion
    
    subroutine test_gravitational()
        real(8), parameter :: G_const = 6.67430d-11
        real(8), parameter :: EARTH_MASS = 5.972d24
        real(8), parameter :: EARTH_RADIUS = 6.371d6
        
        print *, "--- Gravitational Tests ---"
        
        ! Test gravitational force
        call assert_approx(gravitational_force(1000.0d0, 1000.0d0, 1.0d0), &
            G_const * 1.0d6, "gravitational force", 1.0d-7)
        
        ! Test escape velocity (Earth)
        call assert_approx(escape_velocity(EARTH_MASS, EARTH_RADIUS), &
            11186.0d0, "Earth escape velocity", 100.0d0)
        
        ! Test orbital velocity
        call assert_approx(orbital_velocity(EARTH_MASS, EARTH_RADIUS), &
            7909.0d0, "Earth orbital velocity", 100.0d0)
    end subroutine test_gravitational

end program physics_utils_test